build(aio): add grouping of API overloads

petebacondarwin opened the pull request.

petebacondarwin committed:
build(aio): truncate API overview parameters at one line

petebacondarwin committed:
build(aio): render grouped overloads

mary-poppins commented:
You can preview 6f9aee0 at https://pr18081-6f9aee0.ngbuilds.io/.

gkalpak commented on aio/src/styles/2-modules/_details.scss:
```diff
@@ -0,0 +1,32 @@
+  padding: 16px 23px;
+  margin: 0;
+  background-color: $mist;
+  color: #373E41;
```
Hard-coded color 😱

petebacondarwin replied:
I copied this from the resources page styling :-P

gkalpak commented on aio/src/styles/2-modules/_details.scss:
```diff
@@ -0,0 +1,32 @@
+  margin: 0;
+  background-color: $mist;
+  color: #373E41;
+}
```
Maybe also `cursor: pointer`?

gkalpak commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -9,20 +9,20 @@
+    {$ member.name $}{$ member.typeParameters | escape $}{$ params.paramList(member.parameters, truncate) | trim $}
     {%- if member.isOptional  %}?{% endif -%}
-    {$ params.returnType(member.type) $}
+    {$ params.returnType(member.type) | trim | truncateFirstLine $}
```
Should `truncateFirstLine` be conditional?

gkalpak reviewed (commented):
Looks really nice.

Nit:
One thing that threw me off is truncating a multi-line object that also happens to be the last item before a closing `)`:

```ts
foo(bar: {
  baz: qux
})

// ...becomes...

foo(bar: {...)
```

That `{...)` looks kind of weird. It would be awesome if `truncateFirstLine` could detect that `{` is the last char and put a closing `}` at the end.

petebacondarwin committed:
build(aio): truncate API overview parameters at one line

petebacondarwin committed:
build(aio): fix githubLinks line numbers

The API docs have changed. The line numbers are now found in
`startingLine` and `endingLine` properties rather than the `location`
property, which moved into the `fileInfo` property anyway.

petebacondarwin committed:
build(aio): improve API docs layout and styling

petebacondarwin commented:
@gkalpak I fixed the truncating to close any open brackets.

petebacondarwin commented:
I quietened the styling of the overloads, so that they do not look like a new item.
I have also taken the opportunity to tidy up the API doc layout and styling.
Please do take a look when the preview is ready.

mary-poppins commented:
You can preview 6a21073 at https://pr18081-6a21073.ngbuilds.io/.

petebacondarwin commented:
I just realised that we need to be more clever with overloads if the doc type is an interface or abstract class. See https://pr18081-6a21073.ngbuilds.io/api/animations/AnimationBuilder#build

mary-poppins commented:
You can preview 5352936 at https://pr18081-5352936.ngbuilds.io/.

petebacondarwin committed:
build(aio): upgrade to dgeni-packages 0.20.0-rc.5

This fixes unwanted extra overloads in interfaces and abstract classes
This provides the overload grouping support and fixes API doc line numbers

mary-poppins commented:
You can preview 1f08819 at https://pr18081-1f08819.ngbuilds.io/.

petebacondarwin committed:
build(aio): simplify the primary overload's anchor

The primary link to a member should simply be its name, and not require
the parameter list.

mary-poppins commented:
You can preview 6c7952d at https://pr18081-6c7952d.ngbuilds.io/.

mary-poppins commented:
You can preview c6f4261 at https://pr18081-c6f4261.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-base-package/rendering/truncateCode.js:
```diff
@@ -0,0 +1,49 @@
+
+      const parts = str && str.split && str.split(/\r?\n/);
+      if (parts && parts.length > lines) {
+        return balance(parts[0] + '...', ['{', '(', '['], ['}', ')', ']']);
```
❤️
